[Improvement-5539][Master] Check status of taskInstance from cache

blackberrier opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler. Please review https://dolphinscheduler.apache.org/en-us/community/development/pull-request.html before opening a pull request.-->


## Purpose of the pull request

this closes #5539 

Check status of taskInstance from cache, and the cache refresh task state from database periodically.

## Brief change log

Master get task status from taskInstanceCacheManager instead of accessing database.
In taskInstanceCacheManager, I added a thread  which query NEED_FAULT_TOLERANCE task from database and update the cache. To accelerate database access, I use indexed id from cache and check if its status equals NEED_FAULT_TOLERANCE.

## Verify this pull request

This pull request is code cleanup without any test coverage.

blackberrier committed:
improvement:check status of taskInstance from cache

blackberrier committed:
Merge branch 'apache:dev' into dev

CalvinKirs commented:
@ruanwenjun PTAL.Thx

ruanwenjun commented on dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImpl.java:
```diff
@@ -47,6 +58,33 @@
+     */
+    @Override
+    public void run() {
+        while (Stopper.isRunning()) {
```
You should better not implement like this, if you want to close, you can only change the Stop, but Stop is used in many other places.

ruanwenjun commented on dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImpl.java:
```diff
@@ -47,6 +58,33 @@
+                        taskInstanceCache.put(taskInstanceEntry.getKey(), taskInstance);
+                    }
+                }
+                Thread.sleep(CACHE_REFRESH_TIME_MILLIS);
```
You can use Timer to implement, rather than while(true) sleep

ruanwenjun commented on dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterTaskExecThread.java:
```diff
@@ -149,7 +143,10 @@ public Boolean waitTaskQuit() {
                 }
                 // updateProcessInstance task instance
-                taskInstance = processService.findTaskInstanceById(taskInstance.getId());
+                //taskInstance = processService.findTaskInstanceById(taskInstance.getId());
```
Remove the unused code.

ruanwenjun commented on dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImpl.java:
```diff
@@ -47,6 +58,33 @@
+                    TaskInstance taskInstance = processService.findTaskInstanceById(taskInstanceEntry.getKey());
+                    if (null != taskInstance && taskInstance.getState() == ExecutionStatus.NEED_FAULT_TOLERANCE) {
+                        logger.debug("task {} need fault tolerance, update instance cache", taskInstance.getId());
+                        taskInstanceCache.put(taskInstanceEntry.getKey(), taskInstance);
```
You need to consider the concurrency modification. For example, if there is a thread in `MasterTaskExecThread` remove the instance at the same time, these may cause a memory leak.

blackberrier replied:
Hi, @ruanwenjun I don't quite get the situation.
Assume that in one loop, we are checking the task instance, and at this time, `MasterTaskExecThread` remove it. At this point, 'remove' may mean two level, first is database remove, and we will get a null taskInstance. Second is cache remove, which means taskInstance is finished state. In both level, the 'if' condition is not satisfied.
Or another situation, one task is removed outside the entryset loop, then the task will not enter loop.
I wrote a demo for testing , listed below. In timer thread, the sleep simulate a time wasting opration, and another thread remove the entry while timer thread is doing the opration.
```
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ConcurrentHashMap;

public class ConcurrentHashMapRemoveIteratorTest {

    private static ConcurrentHashMap<String, String> map = new ConcurrentHashMap<>();

    public static void main(String[] args) {
        /*for (int i = 0; i < 10; i++) {
            map.put(i + ",", i + "");
        }*/
        map.put(10 + ",", 10 + "");

        new Timer().scheduleAtFixedRate(
                new TimerTask() {
                    @Override
                    public void run() {
                        System.out.println("new loop");
                        for (Map.Entry<String, String> entry : map.entrySet()) {
                            System.out.println("begin" + entry.getKey() + entry.getValue());
                            try {
                                System.out.println("sleeping");
                                Thread.sleep(5000);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                            System.out.println("end" + entry.getKey() + entry.getValue());
                        }
                        System.out.println("new loop end");
                        System.out.println();

                    }
                }, 1000, 10000
        );

        new Thread(() -> {
            try {
                Thread.sleep(12000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            map.remove(10 + ",");
            //map.put(10 + ",","?");
            System.out.println("removed!");

        }).start();
    }
}

```

The output is 
```
new loop
begin10,10
sleeping
end10,10
new loop end

new loop
begin10,10
sleeping
removed!
end10,10
new loop end

new loop
new loop end
```

Maybe there are something wrong in my situation. Can you point out? Thanks.

ruanwenjun replied:
@blackberrier
This is not an atomic operation.
![image](https://user-images.githubusercontent.com/22415594/120464902-05a14100-c3d0-11eb-8329-ab4d6be3198e.png)
When you remove the task in MasterTaskExecThread, but then at the same time you query from database and add the task to cache.
Right?

blackberrier replied:
@ruanwenjun 
I know what's wrong. I though NEED_FAULT_TOLERANCE  tasks will not be remove in MasterExecThread. After I check, taskInstance.getState().typeIsFinished() include FAILURE NEED_FAULT_TOLERANCE. So you are right.

ruanwenjun replied:
@blackberrier The real process should be that the task might be removed from cache first, then add to cache again, and then it will always exist in cache, this cause memory leak 😄

blackberrier committed:
issue5572 use timer instead of while&sleep; consider concurrent modification

blackberrier committed:
Merge branch 'apache:dev' into dev

blackberrier commented:
@ruanwenjun 
Hi, I have updated my code.
1. use timer and scheduleAtFixedRate instead of while(stopper) and thread.sleep. And add close method as well.
2. add synchronized code segment to handler concurrent modification
3. remove useless code

Can you see the latest code? Review it please, Thanks!

ruanwenjun commented on dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImpl.java:
```diff
@@ -114,6 +143,23 @@ public void cacheTaskInstance(TaskExecuteResponseCommand taskExecuteResponseComm
+    class RefreshTaskInstanceTimerTask extends TimerTask {
+        @Override
+        public void run() {
+            synchronized (lock) {
```
I think you can remove lock, use `computeIfPresent` if present to update the status

blackberrier replied:
cool

ruanwenjun commented on dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImpl.java:
```diff
@@ -57,7 +86,7 @@
     @Override
     public TaskInstance getByTaskInstanceId(Integer taskInstanceId) {
         TaskInstance taskInstance = taskInstanceCache.get(taskInstanceId);
```
It is better to use `computeIfAbsent` here.

blackberrier replied:
I put database access out of `computeIfAbsent`'s `Function` class `apply` method, is this ok?

blackberrier replied:
I put database access out of `computeIfAbsent`'s `Function` class `apply` method, is this ok?

ruanwenjun replied:
@blackberrier I think you can change the entire method to :), 
```
return taskInstanceCache.computeIfAbsent(taskInstanceId, k ->processService.findTaskInstanceById(k));
```

ruanwenjun replied:
@blackberrier You can change the entire method to 
```
return taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
```
And you can add some ut.

blackberrier replied:
@ruanwenjun 
I think the database may waste time or something, and thus block other thread. So I take the  `processService.findTaskInstanceById(k)` out

ruanwenjun replied:
@blackberrier I mean you can change this method to 
```
public TaskInstance getByTaskInstanceId(Integer taskInstanceId) {
return taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
}
```
This is an atomic operation.

blackberrier replied:
I know your meaning. I have checked the comment of the `computeIfAbsent`, it says as follows. So I wonder if the database access operation should take out.

![1622725677097](https://user-images.githubusercontent.com/6926304/120649848-c8ac7b80-c4af-11eb-9513-7805180fb339.jpg)

ruanwenjun replied:
@blackberrier 
If there are two threads execute `getByTaskInstanceId` at the same time, we just want to query from database once, right? The block will only happen on the two thread update on the same key, and the key is not exist, this is expected. 

And you should realize that the below way is not an atomic operation, in some situation, it might cause problem.
```java
TaskInstance taskInstance = taskInstanceCache.get(taskInstanceId);
if (taskInstance == null){
    taskInstance = processService.findTaskInstanceById(taskInstanceId);
    taskInstanceCache.put(taskInstanceId,taskInstance);
}
return taskInstance;
```

blackberrier replied:
@ruanwenjun 
I think the block may happen on two thread on two key, and the two key share the same hashcode, which means they are on the same Linkedlist or Red-black tree, and in `computeIfAbsent` synchronized lock on the head node. 
If we write like this,
```
taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
```
and if the first key access database and wasting time , then the second key have to wait.

If we take database access out, like
```
taskInstance = processService.findTaskInstanceById(taskInstanceId);
TaskInstance finalTaskInstance = taskInstance;
taskInstanceCache.computeIfAbsent(taskInstanceId, k -> finalTaskInstance);
```
the second key need not wait while the first key is accessing database.

I wonder if my thought is reasonable.

ruanwenjun replied:
@blackberrier 

> the block may happen on two thread on two key, and the two key share the same hashcode

Yes, but we can avoid the 1 step.
```
1. taskInstance = processService.findTaskInstanceById(taskInstanceId);
2. TaskInstance finalTaskInstance = taskInstance;
3. taskInstanceCache.computeIfAbsent(taskInstanceId, k -> finalTaskInstance);
```

blackberrier replied:
@ruanwenjun 
I can't quite understand.
```
  public TaskInstance getByTaskInstanceId(Integer taskInstanceId) {
        TaskInstance taskInstance = taskInstanceCache.get(taskInstanceId);
        if (taskInstance == null) {
            1. taskInstance = processService.findTaskInstanceById(taskInstanceId);
            2. TaskInstance finalTaskInstance = taskInstance;
            3. taskInstanceCache.computeIfAbsent(taskInstanceId, k -> finalTaskInstance);
        }
        return taskInstance;
    }
```
If we remove the 1 step, then where can we get the taskInstance?

ruanwenjun replied:
@blackberrier I mean you can use below code
```
public TaskInstance getByTaskInstanceId(Integer taskInstanceId) {
return taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
}
```
The cache is used to reduce database pressure, in your plan, if there are multiple threads query the same taskInstance at the same time, they will all query from database.
Your consideration is that when conflicts occur, multiple thread query will block. I think if conflicts occur, if we don't want to be blocked, we should find a better hash method to reduce conflicts. 

In additional, we can use one line of code to achieve, why use more complex code?

blackberrier replied:
@ruanwenjun 
ok, I have commit latest code

blackberrier committed:
use computeifpresent instead of lock

blackberrier committed:
simplify getByTaskInstanceId function

ruanwenjun commented:
LGTM, can you write some unit tests, the coverage is 0 now.

blackberrier committed:
Merge branch 'apache:dev' into dev

blackberrier committed:
add ut for TaskInstanceCacheManagerImpl; fix bug in TaskInstanceCacheManagerImpl

blackberrier commented:
> LGTM, can you write some unit tests, the coverage is 0 now.

@ruanwenjun 
Hi, I've added some unit tests for `TaskInstanceCacheManagerImpl`. Please review it for me. Thx.
By the way, while I am working on this, I have found a bug in `cacheTaskInstance(TaskExecuteResponseCommand taskExecuteResponseCommand)` method , and I have fixed it.

ruanwenjun commented:
@blackberrier Great, I see your job.

ruanwenjun commented on dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImplTest.java:
```diff
@@ -0,0 +1,160 @@
+package org.apache.dolphinscheduler.server.master.cache.impl;
```
You need to add Apache license header here, you can find the license header in other file.
And when you add an ut file, you need to add the test class in root pom.xml like below:
```java
<include>**/api/service/MonitorServiceTest.java</include>
<include>**/api/service/ProcessDefinitionServiceTest.java</include>
<include>**/api/service/ProcessTaskRelationServiceImplTest.java</include>
<include>**/api/service/TaskDefinitionServiceImplTest.java</include>
```

blackberrier replied:
ok, I've added.

blackberrier committed:
add Apache license header;add test class in root pom

ruanwenjun reviewed (approved):
LGTM

CalvinKirs reviewed (approved):
LGTM.let's wait for the CI test  passed

CalvinKirs commented:
Deeply thanks for your contribution, considering that it's your first contribution, I think we can get deep communication, you can contact me through mail or add WeChat(Kris_Evil), when mail or added, please tell me who you are, I think I can help to familiar with the DolphinScheduler if you meet with problems.
